README.md
# flutter_photo_uploader

Flutter Photo Uploader is a Flutter package which uses the Camera package to take a picture, crop it, and post the data to an endpoint. It's simple and lightweight! You can specify a target width and target height for the image during upload as well. It's currently a work in progress with tests being added in the near future. Functionality is also pretty limited right now but it is very easy to use. Images can only be cropped as squares with additional functionality being added to handle all types of cropping.


## Usage

To use this package, add flutter_photo_uploader as a dependency in your pubsec.yaml file. 
For android, You must have to update minSdkVersion to 21 (or higher). On iOS, lines below have to be added inside ios/Runner/Info.plist in order the access the camera.
```
<key>NSCameraUsageDescription</key>
<string>Explanation on why the camera access is needed.</string>
```

## Example
```
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/material.dart';
import 'package:flutter/src/widgets/basic.dart';
import 'package:photo_uploader/photo_uploader.dart';
import 'package:photo_uploader/upload_helper.dart';

void main() {
  // can be called before `runApp()`
  WidgetsFlutterBinding.ensureInitialized();
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: FirstPage(),
    );
  }
}

class FirstPage extends StatefulWidget {
  _FirstPageState createState() => _FirstPageState();
}

class _FirstPageState extends State<FirstPage> {
  UploadHelper helper = new UploadHelper();
  Future<bool> storedFuture;

  @override
  void initState() {
    super.initState();
    storedFuture = helper.available();
  }

  Future upload(ui.Image image) async {
    Uint8List bytes = await helper.getPngByteData(image: image);
    var response = await helper.uploadBytes(
        url: 'https://postman-echo.com/post', bytes: bytes, targetWidth: 100, targetHeight: 100);
    print(response);
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
        future: storedFuture,
        builder: (context, snapshot) {
          // if we have access to cameras, show TakePictureScreen widget
          if (snapshot.hasData && snapshot.data == true) {
            return TakePictureScreen(onUpload: upload);
          } else {
            return Container(
              height: MediaQuery.of(context).size.height,
              width: MediaQuery.of(context).size.height,
              child: Center(
                  child: Text(
                'Cameras not found.',
                style: TextStyle(color: Colors.white, fontSize: 24),
              )),
            );
          }
        });
  }
}
```

## Issues

Please email any issues, bugs, or additional features you would like to see built to arindone@nubeer.io.

## Contributing

If you wish to contribute to this package you may fork the repository and make a pull request to this repository.

CHANGELOG.md
# Changelog

## 0.3.3
- Adding image picker dependency and checking for access to media

## 0.3.2
- Adding loading spinner to indicate image is uploading

## 0.3.1
- updating meta dependency version

## 0.3.0
- Crop screen now starts as full width for crop
- Added targetHeight and targetWidth to make images get resized prior to upload function being called
- Added dispose() to various calls on images to prevent memory leaks

## 0.2.0
- Adding helper function to wrap availableCameras() to UploadHelper class

## 0.1.2
- Updating dependencies

## 0.1.1
- Updating format of files with dartfmt

## 0.1.0
Initial Release
- Delivering an easy to use widget that utilizes the camera package and http package to take, crop, and upload an image to an endpoint


test/photo_uploader_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:photo_uploader/photo_uploader.dart';

void main() async {}


lib/crop_preview.dart
import 'package:flutter/material.dart';

class CropPreview extends StatelessWidget {
  const CropPreview({Key key, this.image}) : super(key: key);
  final image;

  @override
  Widget build(BuildContext context) {
    return SafeArea(
        child: Column(
      children: <Widget>[
        Expanded(
          flex: 6,
          child: FittedBox(
            fit: BoxFit.contain,
            child: SizedBox(
              width: image.width.toDouble(),
              height: image.height.toDouble(),
              child: Stack(children: <Widget>[
                Container(
                  child: FittedBox(
                    fit: BoxFit.contain,
                    child: SizedBox(
                      width: image.width.toDouble(),
                      height: image.height.toDouble(),
                      child: CustomPaint(
                        painter: CropPainter(image),
                        child: Container(),
                      ),
                    ),
                  ),
                )
              ]),
            ),
          ),
        ),
        Container(
          height: 60,
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              ClipOval(
                child: Material(
                  color: Colors.transparent,
                  child: InkWell(
                    splashColor: Colors.grey[500], // inkwell color
                    child: Container(
                      padding: EdgeInsets.all(2.0),
                      decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(100),
                          border:
                              Border.all(width: 1, color: Colors.grey[500])),
                      child: Icon(Icons.arrow_back,
                          color: Colors.white, size: 28.0),
                    ),
                    onTap: () async {
                      Navigator.pop(context);
                    },
                  ),
                ),
              ),
              SizedBox(
                width: 30,
              ),
              ClipOval(
                child: Material(
                  color: Colors.transparent,
                  child: InkWell(
                    splashColor: Colors.grey[500], // inkwell color
                    child: Container(
                      padding: EdgeInsets.all(2.0),
                      decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(100),
                          border:
                              Border.all(width: 1, color: Colors.grey[500])),
                      child: Icon(Icons.cloud_upload,
                          color: Colors.white, size: 28.0),
                    ),
                    onTap: () async {
                      // pop the nav with the image selected
                      Navigator.pop(context, image);
                    },
                  ),
                ),
              ),
            ],
          ),
        )
      ],
    ));
  }
}

class CropPainter extends CustomPainter {
  CropPainter(this.image);
  final image;

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawImage(image, Offset(0.0, 0.0), Paint());
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


lib/upload_helper.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import 'dart:ui' as ui;
import 'package:camera/camera.dart' as cam;
import 'package:meta/meta.dart';

class UploadHelper extends http.BaseClient {
  Map<String, String> requestHeaders;

  static UploadHelper _instance;

  factory UploadHelper() => _instance ??= new UploadHelper._();

  UploadHelper._();

  final client = http.Client();

  Map get headers {
    return this.requestHeaders;
  }

  set headers(Map values) {
    this.requestHeaders = values;
  }

  // wrapper for camera available cameras method
  Future<bool> available() async {
    return await cam.availableCameras().then((value) {
      // return true if we have a cameras array with camera otherwise false
      return value != null && value.length > 0 ? true : false;
    }).catchError((err) {
      return false;
    });
  }

  void resetHeaders() {
    this.headers = this.headers = new Map<String, String>();
  }

  // add a header key value pair
  void addHeader(String key, String value) {
    // if headers is null, set it to an the string and value
    if (this.headers == null) {
      this.headers = new Map<String, String>();
    }

    Map tempHeaders = this.headers;

    // update temp headers with key value
    tempHeaders.update(
      key,
      (existingValue) => value,
      ifAbsent: () => value,
    );

    // set headers from temp headers
    this.headers = tempHeaders;
  }

  Future<dynamic> uploadJSON({String url, String byteString}) async {
    if (byteString == null) {
      throw Exception('ByteString missing for image');
    }
    if (url == null) {
      throw Exception('Url missing for request');
    }

    Map body = new Map<String, String>();
    body['image_data'] = byteString;
    final response = await this.client.post(url,
        headers: this.headers, body: jsonEncode(body), encoding: Utf8Codec());
    if (response.statusCode == 200) {
      // If the server did return a 200 OK response then parse the JSON.
      return jsonDecode(response.body);
    } else {
      // If the server did not return a 200 OK response then throw an exception.
      throw Exception('Failed to load album');
    }
  }

  Future<dynamic> uploadBytes(
      {@required String url,
      @required Uint8List bytes,
      targetWidth = 600,
      targetHeight = 600}) async {
    // this is where we can cap the size to be uploaded
    var imageCodec = await ui.instantiateImageCodec(bytes,
        targetHeight: targetHeight, targetWidth: targetWidth);

    // get the next frame of the image (should be the new size now specified from above)
    ui.FrameInfo frameInfo = await imageCodec.getNextFrame();

    // get the image from the frame info
    ui.Image newImage = frameInfo.image;

    // get byte data from the new image and give it a png format
    ByteData newByteData = await newImage.toByteData(
      format: ui.ImageByteFormat.png,
    );

    // take the byte data and make it a Uint8List
    Uint8List newByteList = newByteData.buffer.asUint8List();

    http.MultipartRequest request =
        http.MultipartRequest('POST', Uri.parse(url));
    http.MultipartFile file = http.MultipartFile.fromBytes(
      'image_data',
      newByteList,
      filename: 'image.png',
      contentType: new MediaType('image', 'png'),
    );

    request.files.add(file);

    // if headers aren't null, use them
    if (this.headers != null) request.headers.addAll(this.headers);

    http.Response response;

    try {
      http.StreamedResponse streamedResponse = await request.send();

      response = await http.Response.fromStream(streamedResponse);
    } catch (e) {
      print(e);
    } finally {
      // release the resource
      newImage.dispose();
    }

    return response;
  }

  Future<String> getBase64PngByteData({ui.Image image}) async {
    ByteData byteData = await image.toByteData(format: ui.ImageByteFormat.png);
    Uint8List pngBytes = byteData.buffer.asUint8List();
    return base64Encode(pngBytes);
  }

  Future<Uint8List> getPngByteData({ui.Image image}) async {
    ByteData byteData = await image.toByteData(format: ui.ImageByteFormat.png);
    Uint8List pngBytes = byteData.buffer.asUint8List();
    return pngBytes;
  }

  Future<http.StreamedResponse> send(http.BaseRequest request) {
    return this.client.send(request);
  }
}


example/lib/main.dart
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/material.dart';
import 'package:photo_uploader/photo_uploader.dart';
import 'package:photo_uploader/upload_helper.dart';
import 'dart:convert';

void main() {
  // can be called before `runApp()`
  WidgetsFlutterBinding.ensureInitialized();
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: FirstPage(),
    );
  }
}

class FirstPage extends StatefulWidget {
  _FirstPageState createState() => _FirstPageState();
}

class _FirstPageState extends State<FirstPage> {
  UploadHelper helper = new UploadHelper();

  Future<bool> storedFuture;

  @override
  void initState() {
    super.initState();
    storedFuture = helper.available();
  }

  // a future to pass in as the upload function
  Future upload(ui.Image image) async {
    try {
      Uint8List bytes = await helper.getPngByteData(image: image);

      var response = await helper.uploadBytes(
          url: 'https://postman-echo.com/post', bytes: bytes, targetWidth: 100, targetHeight: 100);

      Map<String, dynamic> decodedResponse = jsonDecode(response.body);
      print(decodedResponse);

      image.dispose();

    } catch(e) {
      print(e);
    }
  }

  // use a future to make sure we have access to camera before showing a screen
  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
        future: storedFuture,
        builder: (context, snapshot) {
          List<Widget> children;
          // if we have access to cameras, show TakePictureScreen widget
          if (snapshot.hasData && snapshot.data) {
            return TakePictureScreen(onUpload: upload);
          } else if (snapshot.hasError) {
            // if we got an error when trying to check cameras
            children = <Widget>[
              Icon(
                Icons.error_outline,
                color: Colors.red,
                size: 60,
              ),
              Padding(
                padding: const EdgeInsets.only(top: 16),
                child: Text('Error: ${snapshot.error}'),
              )
            ];
          } else if (snapshot.hasData && !snapshot.data) {
            children = <Widget>[
              Text(
                'Cameras not found.',
                style: TextStyle(color: Colors.white, fontSize: 24),
              ),
            ];
          } else {
            children = <Widget>[
              SizedBox(
                child: CircularProgressIndicator(),
                width: 60,
                height: 60,
              ),
              const Padding(
                padding: EdgeInsets.only(top: 16),
                child: Text('Awaiting result...'),
              )
            ];
          }
          // default return either cameras not found, error, or loading
          return Container(
            height: MediaQuery.of(context).size.height,
            width: MediaQuery.of(context).size.height,
            child: Column(
              children: children,
            ),
          );
        });
  }
}
